core/src/Streamly/Internal/Control/Exception.hs

```diff
@@ -103,6 +106,85 @@ data Priority = Priority1 | Priority2 deriving Show
 newtype AllocateIO = AllocateIO
     (forall b c. Priority -> IO b -> (b -> IO c) -> IO (b, IO ()))
 
+allocator :: MonadIO m =>
+       IORef (Int, IntMap (IO ()), IntMap (IO ()))
+    -> Priority
+    -> IO a
+    -> (a -> IO b)
+    -> m (a, m ())
+allocator ref pri alloc free = do
+    let insertResource r (i, mp1, mp2) =
+            case pri of
+                Priority1 ->
+                    ((i + 1, Map.insert i (void $ free r) mp1, mp2), i)
+                Priority2 ->
+                    ((i + 1, mp1, Map.insert i (void $ free r) mp2), i)
+
+    (r, index) <-
+        liftIO $ mask_ $ do
+            -- tid <- myThreadId
+            r <- alloc
+            idx <- atomicModifyIORef' ref (insertResource r)
+            -- liftIO $ putStrLn $ "insert: " ++ show pri
+            --      ++ " " ++ show idx ++ " " ++ show tid
+            return (r, idx)
+
+    let deleteResource (i, mp1, mp2) =
+            case pri of
+                Priority1 ->
+                    let res = Map.lookup index mp1
+                     in ((i, Map.delete index mp1, mp2), res)
+                Priority2 ->
+                    let res = Map.lookup index mp2
+                     in ((i, mp1, Map.delete index mp2), res)
+
+        release =
+            -- IMPORTANT: do not use interruptible operations in this
+            -- critical section. Even putStrLn can make tests fail.
+            liftIO $ mask_ $ do
+                -- tid <- myThreadId
+                -- liftIO $ putStrLn $ "releasing index: " ++ show index
+                --      ++ " " ++ show tid
+                f <- atomicModifyIORef' ref deleteResource
+                -- restoring exceptions makes it non-atomic, tests fail.
+                -- Can use allowInterrupt in "free" if desired.
+                sequence_ f
+    return (r, release)
+
+-- XXX can we ensure via GC that the resources that we are freeing are all
+-- dead, there are no other references to them?
+
+-- | We ensure that all async workers for concurrent streams are stopped
+-- before we release the resources so that nobody could be using the
+-- resource after they are freed.
+--
+-- The only other possibility, could be user issued forkIO not being
+-- tracked by us, however, that would be a programming error and any such
+-- threads could misbehave if we freed the resources from under them.
+--
+-- We use GC based hooks in Stream.bracketIO' so there could be async threads
+-- spawned by GC, releasing resources concurrently with us. For that reason we
+-- need to make sure that the "release" in the bracket end action is executed
+-- only once in that case.
+--
```

Comment:
Is it possible to remove GC hooks? Now that we have prompt cleanup why do we still have them?

Reply:
to cleanup even before the "prompt" release happens. prompt release can be later than GC.